### ISSUE TITLE
Generation of toy data (pdf.create_sampler)

### ISSUE THREAD

**Issue by chm-ipmu** [zfit/zfit#199]
Versions:
python 3.6.9
zfit 0.3.7

I am trying to generate toy data from a PDF, using the create_sampler method.

I have successfully tested  creating a shape from a zfit.pdf.Gauss with no issues.

However, when I try generating from an exponential (also I tested this as failing for Chebyshev) using the following code:

    obs = zfit.Space("obs", limits=(0, 10))
    lambda_ = zfit.Parameter("lambda_", -0.001)
    model = zfit.pdf.Exponential(lambda_, obs)
    sampler = model.create_sampler(n=1000, fixed_params=True)
    sampler.resample()

I get the following error:

Traceback (most recent call last):
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/framework/ops.py", line 1610, in _create_c_op
    c_op = c_api.TF_FinishOperation(op_desc)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Inconsistent values for attr 'T' [DT_INT32, DT_INT32, DT_INT32, DT_VARIANT, DT_INT32, DT_INT32, DT_DOUBLE, DT_STRING, DT_DOUBLE, DT_DOUBLE] vs. [DT_INT32, DT_INT32, DT_INT32, DT_VARIANT, DT_INT32, DT_INT32, DT_DOUBLE, DT_STRING, DT_DOUBLE, DT_RESOURCE] while building NodeDef 'create_sampler/while_sample_data_holder_0' using Op<name=While; signature=input: -> output:; attr=T:list(type),min=0; attr=cond:func; attr=body:func; attr=output_shapes:list(shape),default=[]; attr=parallel_iterations:int,default=10; is_stateful=true>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/gpfs/home/belle2/murphyco/PycharmProjects/flavortaggingvalidation/Fitting/argus_test.py", line 22, in <module>
    toy_argus()
  File "/gpfs/home/belle2/murphyco/PycharmProjects/flavortaggingvalidation/Fitting/argus_test.py", line 10, in toy_argus
    sampler = model.create_sampler(n=1000, fixed_params=True)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/zfit/core/basemodel.py", line 825, in create_sampler
    name=name)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/zfit/core/data.py", line 606, in from_sample
    name="sample_data_holder_{}".format(cls.get_cache_counting()))
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 258, in __call__
    return cls._variable_v1_call(*args, **kwargs)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 219, in _variable_v1_call
    shape=shape)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 197, in <lambda>
    previous_getter = lambda **kwargs: default_variable_creator(None, **kwargs)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variable_scope.py", line 2507, in default_variable_creator
    shape=shape)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 262, in __call__
    return super(VariableMetaclass, cls).__call__(*args, **kwargs)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/resource_variable_ops.py", line 1406, in __init__
    distribute_strategy=distribute_strategy)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/resource_variable_ops.py", line 1579, in _init_from_args
    initial_value),
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 2711, in _try_guard_against_uninitialized_dependencies
    return _safe_initial_value_from_tensor(name, initial_value, op_cache={})
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 2751, in _safe_initial_value_from_tensor
    new_op = _safe_initial_value_from_op(name, op, op_cache)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 2786, in _safe_initial_value_from_op
    new_op_input = _safe_initial_value_from_tensor(name, op_input, op_cache)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 2751, in _safe_initial_value_from_tensor
    new_op = _safe_initial_value_from_op(name, op, op_cache)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 2786, in _safe_initial_value_from_op
    new_op_input = _safe_initial_value_from_tensor(name, op_input, op_cache)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 2751, in _safe_initial_value_from_tensor
    new_op = _safe_initial_value_from_op(name, op, op_cache)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 2786, in _safe_initial_value_from_op
    new_op_input = _safe_initial_value_from_tensor(name, op_input, op_cache)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 2751, in _safe_initial_value_from_tensor
    new_op = _safe_initial_value_from_op(name, op, op_cache)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/ops/variables.py", line 2802, in _safe_initial_value_from_op
    attrs=op.node_def.attr)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/framework/ops.py", line 3360, in create_op
    attrs, op_def, compute_device)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/framework/ops.py", line 3429, in _create_op_internal
    op_def=op_def)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/framework/ops.py", line 1773, in __init__
    control_input_ops)
  File "/home/belle2/murphyco/miniconda3/envs/jeta/lib/python3.6/site-packages/tensorflow_core/python/framework/ops.py", line 1613, in _create_c_op
    raise ValueError(str(e))
ValueError: Inconsistent values for attr 'T' [DT_INT32, DT_INT32, DT_INT32, DT_VARIANT, DT_INT32, DT_INT32, DT_DOUBLE, DT_STRING, DT_DOUBLE, DT_DOUBLE] vs. [DT_INT32, DT_INT32, DT_INT32, DT_VARIANT, DT_INT32, DT_INT32, DT_DOUBLE, DT_STRING, DT_DOUBLE, DT_RESOURCE] while building NodeDef 'create_sampler/while_sample_data_holder_0' using Op<name=While; signature=input: -> output:; attr=T:list(type),min=0; attr=cond:func; attr=body:func; attr=output_shapes:list(shape),default=[]; attr=parallel_iterations:int,default=10; is_stateful=true>

Is this a misunderstanding on my part, or a bug?
Cheers!

**Comment by mayou36** [comment]
Hm, that looks odd.

Can you maybe upgrade to the newest version, 0.4.1 and see if the issue persists?

**Comment by chm-ipmu** [comment]
Dear Jonas,

I tried to do that afterwards, as I saw a new version was available on 
the docs page.

I tried installing using conda (with no other dependencies) but was 
still given version 0.3.7.

Is the v4 release available from conda-forge?

**Comment by chrisburr** [comment]
Can you try forcing the version to update `conda install "zfit>=0.4.1"`? If you run in to conflicts it might be easiest to make a new environment:

```bash
conda create --name updated-zfit python=3.6 zfit # + any other packages you need
```

If that still doesn't work can you share the error messages you get as well as the output of `conda info --all` and `conda config --show`.

**Comment by chm-ipmu** [comment]
It appears that upgrading to the latest version has solved the issue.

You may be interested that I initially tried creating an environment via the use on `conda env create` with the only listed dependency as zfit, but this did not install the latest version, rather it installed 3.6.0. I needed to force the latest version using the > sign.

Am I right in thinking that now I have upgraded to v4 I must change all dependent parameters from using standard python operators to function calls?

This will be quite cumbersome for me. Is there any way to turn off eager evaluation or otherwise to retain the use of code like `shifted_mean = mean + shift` ?

**Comment by mayou36** [comment]
>You may be interested that I initially tried creating an environment via the use on conda env create with the only listed dependency as zfit, but this did not install the latest version, rather it installed 3.6.0. I needed to force the latest version using the > sign.

Hm, I just tested that for myself. I have the `conda-forge` channel added with preference, nothing else. For me, it grabs the 0.4.1. Maybe it has to do with TF dependencies or something. Maybe upgrade conda?

> Am I right in thinking that now I have upgraded to v4 I must change all dependent parameters from using standard python operators to function calls?

"yes": all you need to do is to have a function instead of a simple value.

e.g. your example:
`shifted_mean = mean + shift`
should now be:
`shifted_mean = lambda: mean + shift`
and give this to the CompositeParameter.
(and you need to give the dependents to the CompositeParameter, in this case
I assume `shift` is a parameter, `mean` is fixed. But you should also be able to give both)

> This will be quite cumbersome for me. 

I can quite imagine, we have a lot of codebase (not just zfit) that had to be changed as well. But I see it from the positive side: the changes in TF2 (if you're aware of them) are actually _huge_. We were able to catch and hide many of this things by working hard behind the scenes and this this with the `CompositeParameter` is one of the very few things that is technically not hideable.

Furthermore, it is actually the better way: it's more explicit. Think of it as if we were using Numpy. And we _could_ have "guessed that already in the very beginning, but designing a library is hard, something turn out to be genious ideas and others to be simply dumb and wrong; in hindsight of course ;)
Also, it is possible that these kind of changes happen again until the 1.0 release. But, as also these changes, they will be minor and we try to affect as few people as possible (it's a normality with other libraries as well, e.g. `Pandas` recent 1.0 release etc.). We will however always communicate all changes as clearly as possible, make upgrade guides and provide as much help as we can.

If you have ideas or feedback for the design of the library or the behavior, you're also very welcome to propose things; in the end this is the only way we can develop this further.

> Is there any way to turn off eager evaluation or otherwise to retain the use of code like `shifted_mean = mean + shift` ?

No, intentionally. It's better to break hard and make things the better way. It's some work, but then it's done.

And my personal note on this: It took me a long way to accept and realize that things are better the eager way. But now, in the end, I am fully convinced that this is the right way to go, also in terms of other backends etc.

If you wanna e.g. an in-depth discussion of us (and our initial position on the graph mode) with some Google-guys: https://groups.google.com/a/tensorflow.org/forum/#!topic/discuss/yaYid8h85s0

Let me know if there are any other issues

**Comment by marinang** [comment]
`ComposedParameter` instead `CompositeParameter` :)

**Comment by mayou36** [comment]
Hi @chm-ipmu, is this actually resolved I assume? Can you maybe confirm?

**Comment by chm-ipmu** [comment]
Sorry - yes the problem resolved itself after updating to zfit 0.4. Cheers
